packages/oidc-dpop-client-browser/src/dpop/tokenExchange.ts

```diff
@@ -0,0 +1,206 @@
+  sub: string;
+  iss: string;
+  webid?: string;
+};
+
+function isWebidOidcIdToken(
+  token: WebidOidcIdToken | Record<string, unknown>
+): token is WebidOidcIdToken {
+  return (
+    (token.sub &&
+      typeof token.sub === "string" &&
+      token.iss &&
+      typeof token.iss === "string" &&
+      !token.webid) ||
+    typeof token.webid === "string"
+  );
+}
+
+/**
+ * Extracts a Webid from an ID token based on https://github.com/solid/webid-oidc-spec.
+ * The upcoming spec is still a work in progress.
+ *
+ * Note: this function does not implement the userinfo webid lookup yet.
+ * @param idToken
+ */
+async function deriveWebidFromIdToken(idToken: string): Promise<string> {
+  const decoded = await decodeJwt(idToken);
+  if (!isWebidOidcIdToken(decoded)) {
+    throw new Error(
+      `Invalid ID token: ${JSON.stringify(
+        decoded
+      )} is missing 'sub' or 'iss' claims`
+    );
+  }
+  if (decoded.webid) {
+    return decoded.webid;
+  }
+  if (!decoded.sub.match(/^https?:\/\/.+\..+$/)) {
+    throw new Error(
+      `Cannot extract WebID from ID token: the ID token returned by ${decoded.iss} has no webid claim, nor an IRI-like sub claim: [${decoded.sub}]`
+    );
+  }
+  return decoded.sub;
+}
+
+function validatePreconditions(
+  issuer: IIssuerConfig,
+  data: TokenEndpointInput
+): void {
+  if (
+    data.grantType &&
+    (!issuer.grantTypesSupported ||
+      !issuer.grantTypesSupported.includes(data.grantType))
+  ) {
+    throw new Error(
+      `The issuer [${issuer.issuer.toString()}] does not support the [${
+        data.grantType
+      }] grant`
+    );
+  }
+  if (!issuer.tokenEndpoint) {
+    throw new Error(
+      `This issuer [${issuer.issuer.toString()}] does not have a token endpoint`
+    );
+  }
+}
+
+function validateTokenEndpointResponse(
+  tokenResponse: Record<string, unknown>
+): Record<string, unknown> & { access_token: string; id_token: string } {
+  if (!hasAccessToken(tokenResponse)) {
+    throw new Error(
+      `Invalid token endpoint response: ${JSON.stringify(
+        tokenResponse
+      )} is missing an access_token.`
+    );
+  }
+
+  if (!hasIdToken(tokenResponse)) {
+    throw new Error(
+      `Invalid token endpoint response: ${JSON.stringify(
+        tokenResponse
+      )} is missing an id_token.`
+    );
+  }
+  return tokenResponse;
+}
+
+export async function getTokens(
+  issuer: IIssuerConfig,
+  client: IClient,
+  data: TokenEndpointInput,
+  dpop: boolean
+): Promise<TokenEndpointResponse | undefined> {
+  validatePreconditions(issuer, data);
+  const headers: Record<string, string> = {
+    "content-type": "application/x-www-form-urlencoded",
+  };
+  let dpopJwk: JSONWebKey | undefined = undefined;
+  if (dpop) {
+    dpopJwk = await generateJwkForDpop();
+    headers["DPoP"] = await createDpopHeader(
+      issuer.tokenEndpoint,
+      "POST",
+      dpopJwk
+    );
+  }
+  // TODO: is this necessary ? it's present in OAuth2 in action book, but not in spec
```

Comment:
Are we following up on this anywhere? If so, maybe log that somewhere public and link to that issue here.

Reply:
There isn't any tickets dedicated to this per se, I wanted to test it as part of SDK-1342 (which would enable end-to-end testing of the DPoP features). I preferred to keep it as a comment to avoid forgetting about it and searching around why requests are failing in case it is actually needed ^^

Reply:
OK, if you're going to get to it in due time, then that's fine I think.

Reply:
Depending on how the client was registered (static/dynamic), the token endpoint may require authentication via headers (as shown here in the commented-out code), or authentication via form params, or no authentication at all. The client would presumably know which mode of operation is in place.

Reply:
Interesting. @acoburn - would client registration (i.e. static or dynamic) **_necessarily_** have any bearing on whether the token endpoint requires auth or not? (i.e. I would have thought the server may, or may not, require auth on any endpoint it offers based on a whatever criteria it wants, e.g. always required, never required, only required at certain times of the day, required for requests from certain incoming IP ranges, only required for dynamically registered clients, only required for statically registered clients, etc. etc.)
I assume there's no spec-defined way for a client to 'discover' at runtime whether any endpoint our client is calling requires auth? I'm guessing the book is just showing a simple example of where a server does happen to require Basic auth on it's token endpoint, but since our library can't really know beforehand (and can't determine at runtime either) whether the authorization server we're hitting **_requires_** it or not, I kinda think "just adding it anyway" can't really cause any harm (i.e. if the server doesn't require it, then it'll just ignore it (I think!)), and if it does happen to require it (and it expects the value to be "clientId:clientSecret", which seems reasonable to assume), then we're well-and-good too! So does anyone see any problems with "just adding it anyway" if we have a client secret (i.e. just uncommenting this code)...?